### ISSUE TITLE
WPF using Python Machine Learning Libraries

### ISSUE THREAD

**Issue by redcican** [pythonnet/pythonnet#944]
Hi, I was just wondering if I am in the right place?
Currently, I am working on a project to create a GUI by using WPF, we want to utilize the majority of Python Machine Leaning Libraries (Scikit-Learn, TensorFlow, Pandas, Numpy). So my question is: can I take advantage from PythonNet?

**Comment by lostmsu** [comment]
@redcican not sure about Pandas and Scikit, but TensorFlow and NumPy should definitely work. In fact, my company makes statically-typed wrapper for TensorFlow using Python.NET: https://losttech.software/gradient.html

**Comment by redcican** [comment]
@lostmsu Thanks, so your company used PythonNet as the backend to support the communicate between C# and Python?

**Comment by lostmsu** [comment]
@redcican yes, correct.

**Comment by redcican** [comment]
@lostmsu Do you have any idea, if it is possible to utilize PythonNet for WPF?

**Comment by lostmsu** [comment]
@redcican  which way do you want it? WPF app calling into Python code, or Python app showing WPF UI? Calling into Python code from WPF would be easy. Using WPF from Python is a bit more tricky, because WPF apps normally start using WPF-specific startup code.

**Comment by redcican** [comment]
I want to use WPF to call Python code. Say we have an option to build Neural Network in C#. I want to give the user button so they can change the architecture of Neural Network. Something like that! What do you mean easy? 🤣

**Comment by lostmsu** [comment]
@redcican this example is literally in repo's README:

```csharp
using (Py.GIL())
{
    dynamic np = Py.Import("numpy");
    Console.WriteLine(np.cos(np.pi * 2));
}
```

You just need to ensure the right version of Python is installed with NumPy.

**Comment by redcican** [comment]
@lostmsu I have tried this code before, but I don't know it works also for more complicated scenarios. Like event, delegate and so on. Anyway, I will give a try and give you feedback! Thank you very much.

**Comment by lostmsu** [comment]
@redcican if you pass in delegates, you will have to wrap their bodies in `using Py.Gil()` too

**Comment by redcican** [comment]
@lostmsu I just tried to install `Gradient` into my computer. But I got an error
```
 Could not find a version that satisfies the requirement tensorflow==1.10.* (from versions: 1.13.0rc1, 1.13.0rc2, 1.13.1, 1.13.2, 1.14.0rc0, 1.14.0rc1, 1.14.0, 2.0.0a0, 2.0.0b0, 2.0.0b1)
No matching distribution found for tensorflow==1.10.*```

Will you update your dependencies?

**Comment by lostmsu** [comment]
@redcican that tells you you have to install 1.10, or explicitly use one of `GradientSetup.UseEnvironment*` methods.

**Comment by redcican** [comment]
@lostmsu I followed this instruction 
```
Install TensorFlow 1.10 using pip:

Find python executable in the installation directory (VS installs to: C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python36_64\python.exe)

Open command line to the directory, containing python

Execute .\python -m pip install "tensorflow==1.10.*" or .\python -m pip install "tensorflow-gpu==1.10.*" if you want GPU acceleration via CUDA (NVidia only, also requires matching CuDNN and CUDA, see instructions)
```
But it doesn't work!

By the way, I have Python 3.7? Will it be the problem?

**Comment by lostmsu** [comment]
@redcican what is the error you are seeing?

**Comment by redcican** [comment]
I just showed you before.

```
 Could not find a version that satisfies the requirement tensorflow==1.10.* (from versions: 1.13.0rc1, 1.13.0rc2, 1.13.1, 1.13.2, 1.14.0rc0, 1.14.0rc1, 1.14.0, 2.0.0a0, 2.0.0b0, 2.0.0b1)
No matching distribution found for tensorflow==1.10.*```

Maybe Google doesn't support anymore for tensorflow 1.10?

**Comment by lostmsu** [comment]
@redcican there's something weird with your Python installation. This is what I get trying to install non-existing version of TensorFlow. Perhaps your company has its own pip repository?

> python -m pip install "tensorflow=="                                                   
> Collecting tensorflow==                                                                                     
> ERROR: Could not find a version that satisfies the requirement tensorflow== (from versions: 1.2.0rc2, 1.2.0, 1.2.1, 1.3.0rc0, 1.3.0rc1, 1.3.0rc2, 1.3.0, 1.4.0rc0, 1.4.0rc1, 1.4.0, 1.5.0rc0, 1.5.0rc1, 1.5.0, 1.5.1, 1.6.0rc0, 1.6.0rc1, 1.6.0, 1.7.0rc0, 1.7.0rc1, 1.7.0, 1.7.1, 1.8.0rc0, 1.8.0rc1, 1.8.0, 1.9.0rc0, 1.9.0rc1, 1.9.0rc2, 1.9.0, 1.10.0rc0, 1.10.0rc1, 1.10.0, 1.11.0rc0, 1.11.0rc1, 1.11.0rc2, 1.11.0, 1.12.0rc0, 1.12.0rc1, 1.12.0rc2, 1.12.0, 1.12.2, 1.12.3, 1.13.0rc0, 1.13.0rc1, 1.13.0rc2, 1.13.1, 1.13.2, 1.14.0rc0, 1.14.0rc1, 1.14.0, 2.0.0a0, 2.0.0b0, 2.0.0b1)
> ERROR: No matching distribution found for tensorflow==

**Comment by lostmsu** [comment]
Either way, you can install 1.13 instead of 1.10. You would just have to configure the environment using one of `GradientSetup.UseEnvironment*`

**Comment by redcican** [comment]
@lostmsu We don't have our own pip repository with only 1.13+. I checked the Website of TensorFlow, they still support for version 1.10. I shall install TensorFlow 1.13. Any I will give you feedback!

Many thanks.

**Comment by lostmsu** [comment]
You're welcome! Though if you have any Gradient-specific questions, it is better to create an issue/question in https://github.com/losttech/Gradient .
For non-Gradient related questions here is a good place.

**Comment by redcican** [comment]
@lostmsu Ok, got it!
